[WIP] ec2_snapshot: enable permissioning

willthames opened the pull request:
##### SUMMARY
Allow createVolumePermission snapshot attributes to be set.

Update tagging to be idempotent and correctly report changes.

Use Ansible's built in parameter validation methods (`required_if`,
`mutually_exclusive`, `required_together`).

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
ec2_snapshot

##### ANSIBLE VERSION
```
ansible 2.3.0 (devel fbf55886ba) last updated 2017/03/08 15:41:54 (GMT +1000)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Jan 12 2017, 17:59:37) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]
```


##### ADDITIONAL INFORMATION

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_snapshot.py:
```diff
@@ -80,6 +80,14 @@
     required: false
     default: 0
     version_added: "2.0"
+  create_volume_permission:
```
should be plural

willthames replied:
Good catch

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_snapshot.py:
```diff
@@ -239,16 +260,40 @@ def create_snapshot(module, ec2, state=None, description=None, wait=None,
+
+        permissions = ec2.get_snapshot_attribute(snapshot.id)
+
+        users_to_add, users_to_remove = existing_to_desired(permissions.attrs.get('user_ids', []),
```
If you are removing the last user, users_to_add == ['None'], which is truthy so it fails on line 280. Same problem with remove the group "all".

Commit:
ec2_snapshot: enable permissioning

Allow createVolumePermission snapshot attributes to be set.

Update tagging to be idempotent and correctly report changes.

Use Ansible's built in parameter validation methods (`required_if`,
`mutually_exclusive`, `required_together`).

Ensure check_mode works correctly

s-hertel commented on lib/ansible/modules/cloud/amazon/ec2_snapshot.py:
```diff
@@ -231,24 +251,66 @@ def create_snapshot(module, ec2, state=None, description=None, wait=None,
                      volume_size=snapshot.volume_size,
-                     tags=snapshot.tags.copy())
+                     tags=tags,
+                     permissions=permissions.attrs)
```
This will fail in check mode if groups are being added or removed since permissions won't have attrs. I also notice there seems to be a lag with updating removals - I can remove a user and it will show the results as being changed but the user id remains. Would it make sense to always set permissions to create_volume_permissions if a change is being made? Although it would be easier to see things going wrong if a call is actually being made to get the latest state...

willthames commented:
Moving to WIP status, will convert to boto3 and add tests

gundalow commented:
@willthames candidate to close?
